Stop checking for RenderStyle::whiteSpace() in TextManipulationController.cpp

nt1m opened the pull request:
#### 4c6946f194a3e7068a4fec0fe216d682951c89c6
<pre>
Stop checking for RenderStyle::whiteSpace() in TextManipulationController.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=258544">https://bugs.webkit.org/show_bug.cgi?id=258544</a>
rdar://111369337

Reviewed by Wenson Hsieh.

Stop checking for white-space-collapse: collapse, and only check for text-wrap: nowrap.
We always want to preserve whitespace regardless if white-space-collapse is discard/collapse/preserve/...

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::isEnclosingItemBoundaryElement):

Canonical link: <a href="https://commits.webkit.org/265568@main">https://commits.webkit.org/265568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c84ed661ec2706d256fb980f42a914329ff28ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11786 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12894 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10723 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13832 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11438 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13632 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9503 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13304 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10179 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17382 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10655 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10334 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13558 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8846 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9933 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2696 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14208 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10617 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -337,7 +337,7 @@ static bool isEnclosingItemBoundaryElement(const Element& element)
 
         // Treat a or li with white-space: nowrap as its own paragraph so that wrapping whitespace between them will be preserved.
-        if (renderer->style().whiteSpace() == WhiteSpace::NoWrap)
+        if (renderer->style().textWrap() == TextWrap::NoWrap && renderer->style().whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse)
```
@whsieh do you know if we actually care about whitespace collapse here? Replacing white-space with text-wrap/collapse in many cases is not a mechanical change but rather a chance to drop one of the checks.
@nt1m please either remove or update the comment above.

whsieh replied:
I reviewed this on the assumption that there would be no behavior change, based on Tim's commit message (as well as the fact that the API test `TextManipulation.CompleteTextManipulationPreservesWhitespacesBetweenItems` still passes).

It would help to test the original use case (webpage translation in https://www.ukr.net/ua/), and verify that that bug remains fixed.

nt1m replied:
It might be possible to drop the `white-space-collapse` check based on the comment, given it seems like `white-space-collapse: preserve` would preserve the wrapping whitespace anyway? But since it shouldn't make a difference, I decided to stay on the safe side.

alanbaradlay replied:
I think reducing code complexity wins over staying on the safe side, when we don't have to rush the change in.

nt1m replied:
OK, I'll give it a shot then.

nt1m committed:
Stop checking for RenderStyle::whiteSpace() in TextManipulationController.cpp
https://bugs.webkit.org/show_bug.cgi?id=258544
rdar://111369337

Reviewed by Wenson Hsieh.

Stop checking for white-space-collapse: collapse, and only check for text-wrap: nowrap.
We always want to preserve whitespace regardless if white-space-collapse is discard/collapse/preserve/...

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::isEnclosingItemBoundaryElement):

Canonical link: https://commits.webkit.org/265568@main
